modules/standard/ChapelVersion.chpl

```diff
@@ -0,0 +1,230 @@
+
+* :proc:`version`: a utility function for creating new version values
+
+* :type:`versionValue`: the record type used to represent version
+  values, supporting compile-time comparisons
+
+*/
+
+module ChapelVersion {
+  // Query the compiler to get its version number and cache those values
+  private param major = __primitive("version major");
+  private param minor = __primitive("version minor");
+  private param update = __primitive("version update");
+  private param SHA = __primitive("version sha");
+
+
+  /* The version of ``chpl`` used to compile this program. */
+  const chplVersion: versionValue(?);
+  chplVersion = new versionValue(major, minor, update, SHA);
+
+  
+  /*
+    A helper function that creates a new version value from its
+    arguments.
+
+    :arg major: The major version number
+    :type major: `int`
+
+    :arg minor: The minor version number
+    :type minor: `int`
+
+    :arg update: The optional update version number (defaults to 0)
+    :type major: `int`
+
+    :arg sha: The optional SHA (defaults to "")
+    :type major: `int`
+
+    :returns: A new version value
+  */
+     
+  proc version(param major: int, param minor: int,
+               param update: int = 0, param sha: string=""): versionValue(?) {
+    return new versionValue(major, minor, update, sha);
+  }
+
+
+  /* This record represents a Chapel version value.  It uses ``param``
+     fields to represent its components in order to support
+     compile-time comparison of version numbers which in turn permits
+     code to specialize to specific versions of Chapel.  Values of
+     :type:`versionValue` are printed out identically to `chpl
+     --version`.
+  */
+  record versionValue {
+    /* The major version number.  For version ``2.0.1``, this would be
+       ``2``. */
+    param major: int;
+
+    /* The minor version number.  For version ``2.0.1``, this would be
+       ``0``. */
+    param minor: int;
+
+    /* The update version number.  For version ``2.0.1``, this would
+       be ``1``. */
+    param update: int;
+
+    /* For pre-releases of Chapel, the git SHA from which the compiler
+       was built.  For official releases, this will be the empty
+       string. */
+    param sha: string;
+
+    /* Indicates whether or not this is a pre-release of Chapel.  This
+       is currently implemented as a method, but may become a field
+       in future versions of this module. */
+    proc preRelease param : bool {
+      return sha != "";
+    }
+
+    /* Indicates whether or not this is an official release of Chapel.
+       This is always the opposite of ``preRelease``. */
+    proc isOfficialRelease() param : bool {
+      return !preRelease;
+    }
+
+    pragma "no doc"
+    proc writeThis(s) throws {
+      s.write("chpl version ", major, ".", minor, ".", update);
+      if preRelease then
+        s.write(" pre-release (", sha, ")");
+    }
+  }
+
+  proc ==(v1: versionValue(?), v2: versionValue(?)) param {
```

Comment:
There's a lot of very similar code for all these comparisons. Looking at the python distutils version implementation [here](https://docs.openstack.org/tooz/latest/_modules/distutils/version.html), I wonder if we could implement a single function that returned `-1 for <`, `0 for =`, and `1 for >` and use that to implement the comparison methods.

On the other hand I suppose what's here is already implemented, so maybe it's not worth the effort.

Reply:
I was thinking that the `compilerError()` would be a challenge to that "spaceship operator"-like approach, but I think that was when I still had some errors in my logic.  I also wouldn't want to call the helper twice for `<` and `=` (which might suggest returning -2, -1, 0, 1, and 2 or something like that?).  Mostly, it was easiest for me to think in these terms once I got the first one correct.

Reply:
Re. "spaceship operator" - one day, I hope we will have the `<==>` operator that can implement all of these at once (see #8545)

Reply:
In a subsequent draft, I used a private routine named `spaceship` to avoid a lot of the repeated code in other cases.  Micro-optimizing, it's interesting to me to see how some queries might be answered more quickly/directly by explicit operator overloads than by a general spaceship operator, though for this case it's not likely to be a factor (given that these are evaluated at compile-time).